app/src/main/java/uk/ac/ebi/atlas/search/JsonGeneSearchController.java

```diff
@@ -46,78 +51,104 @@ public class JsonGeneSearchController extends JsonExceptionHandlingController {
-                            var cellIds = exp2cells.getValue();
-
-                            var experimentAttributes =
-                                    ImmutableMap.<String, Object>builder().putAll(
-                                            getExperimentInformation(experimentAccession, geneId));
-                            var facets =
-                                    ImmutableList.<Map<String, String>>builder().addAll(
-                                            unfoldFacets(geneSearchService.getFacets(cellIds)
-                                                    .getOrDefault(experimentAccession, ImmutableMap.of())));
-
-                            if (markerGeneFacets.containsKey(geneId) &&
-                                    markerGeneFacets.get(geneId).containsKey(experimentAccession)) {
-                                facets.add(
-                                        ImmutableMap.of(
-                                                "group", MARKER_GENE.getTitle(),
-                                                "value", "experiments with marker genes",
-                                                "label", "Experiments with marker genes",
-                                                "description", MARKER_GENE.getTooltip()));
-                                experimentAttributes.put(
-                                        "markerGenes",
-                                        convertMarkerGeneModel(
-                                                experimentAccession,
-                                                geneId,
-                                                markerGeneFacets.get(geneId).get(experimentAccession)));
-                            } else {
-                                experimentAttributes.put(
-                                        "markerGenes", ImmutableList.of());
-                            }
-
-                            return ImmutableMap.of("element", experimentAttributes.build(), "facets", facets.build());
-
-                        })).collect(toImmutableList());
-
-        var matchingGeneIds = "";
-        if (geneIds.get().size() == 1 && !geneIds.get().iterator().next().equals(geneQuery.queryTerm())) {
-            matchingGeneIds = "(" + String.join(", ", geneIds.get()) + ")";
+        var resultsBuilder = ImmutableList.builder();
+        // This was a very expressive stream chain (look it up in the history), but to filter by results where the gene
+        // is a marker, making it a for loop is more readable IMHO; I also noticed a 10% performance improvement
+        for (var i = 0; i < expressedGeneIdEntries.size(); i++) {
+            var geneId = expressedGeneIdEntries.get(i).getKey();
+
+            var exp2cellsIt = expressedGeneIdEntries.get(i).getValue().entrySet().iterator();
+            while (exp2cellsIt.hasNext()) {
+                var exp2cells = exp2cellsIt.next();
+                var experimentAccession = exp2cells.getKey();
```

Comment:
I suggest to use the for-each loop instead of the classic for/while loop, as you don't really need the `i` variable and it is more readable and more importantly!!! ;-) less code to maintain

```suggestion
       for (var geneIdEntry : expressedGeneIdEntries) {
            var geneId = geneIdEntry.getKey();

            for (var exp2cells: geneIdEntry.getValue().entrySet()) {
                var experimentAccession = exp2cells.getKey();
```

Reply:
Totally true. I’m so used to using streams that I had forgotten about the auto-increment notation. :crying_cat_face: